### ISSUE TITLE
arm32 support

### ISSUE THREAD

**Issue by jmkrauz** [bytecodealliance/wasmtime#1173]
- What is the feature or code improvement you would like to do in Cranelift?
I would like to add ARM support by implementing ARM (not Thumb) encodings/recipes and abi. I am especially interested in armv7 little endian support.
- What is the value of adding this in Cranelift?
Emit code that could run on ARM architecture.
- Do you have an implementation plan, and/or ideas for data structures or
  algorithms to use?
At first i want to add basic integer operations followed by floating point instructions. I thought about implementing the latter using VFP rather than the NEON extension (As far as I know NEON does not support double precision). As for calling convention, I would like to implement AAPCS.
 I already have some code that i hope to push after necessary modifications.
- Have you considered alternative implementations? If so, how are they better
  or worse than your proposal?
No, but I am open to suggestions.

**Comment by squillace** [comment]
Hi all; what's the status with ARM32/ARM 64? It appears that #1344 was merged. Love to use wasmtime on Pis..... and I thank you for your efforts here.

**Comment by bjorn3** [comment]
As part of #1174 an ARM backend is being developed. It may take a while before it is complete enough to be merged into master, as #1174 is a big change to Cranelift. Deveopment happens at the https://github.com/cfallin/wasmtime/ fork if you want to try it. Discussion happens at https://bytecodealliance.zulipchat.com/#narrow/stream/225524-cranelift-new-backend

**Comment by toothbrush7777777** [comment]
It appears that AArch64/ARM64 support was merged in #1494.

**Comment by tschneidereit** [comment]
Indeed! \o/

Note that it's not yet entirely complete, and not passing the entire test suite. It's getting close, though!

**Comment by autodidaddict** [comment]
Just dropping in here to check on whether there's a timeframe for when armv7 compilation might be supported? This basically includes 95% of the Raspberry Pis on the market today. aarch64 (as far as I've been able to determine) is only applicable to the Pi4s, and only then if you've managed to install the 64-bit Raspbian.

**Comment by toothbrush7777777** [comment]
@autodidaddict The Raspberry Pi 3B, 3B+ and 4B support AArch64 (ARMv8).

**Comment by autodidaddict** [comment]
When I attempt to run an aarch64 binary on the pi3b+ it tells me it's the wrong file type. Does one need something other than stock Raspian?

**Comment by bjorn3** [comment]
Stock Raspian is 32bit, so it doesn't support 64bit binaries.

**Comment by autodidaddict** [comment]
Ok so it looks like I just need an os like Gentoo that gives me aarch64 userland on the pi3b+. Thanks for the info!

**Comment by toothbrush7777777** [comment]
@autodidaddict You can try the new beta 64-bit version of Raspberry Pi OS (https://www.raspberrypi.org/forums/viewtopic.php?f=117&t=27537) or use Ubuntu or Gentoo.

**Comment by julian-seward1** [comment]
FWIW, I have been running 64-bit Fedora 32 on my Pi 3B for months, maybe even a year, and it works just great.  wasmtime, using the new aarch64 backend under development, works on it no problem.

**Comment by squillace** [comment]
Yup, some issues here and there remain for 32-bit builds:

>
>error[E0308]: mismatched types
>  --> crates/environ/src/tunables.rs:65:34
>   |
> 65 |             static_memory_bound: (10 * (1 << 20)) / crate::WASM_PAGE_SIZE,
 >  |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u64`, found `u32`
>

**Comment by bjorn3** [comment]
Not just that, I believe the arm32 backend is too incomplete to do much useful.

**Comment by cfallin** [comment]
Indeed, the high-order bit is that the compiler backend on arm32 can only generate code for 32-bit operators right now, not 64-bit ones, and in general is pretty incomplete. There is some hope that the new isel framework could make this work easier to complete, but until then, mismatching types in the runtime are the least of our worries :-)

**Comment by akirilov-arm** [comment]
Note (@jmkrauz in particular) that in issue #3721 we are discussing potential removal of the 32-bit Arm backend.

**Comment by cfallin** [comment]
For completeness to keep this issue up-to-date: we indeed removed the halfway-complete ARM32 backend because, unfortunately, no one had the resources to bring it up to a Wasm-MVP level of completeness, nor maintain it as we continue to evolve the compiler. However, we would absolutely welcome a new contribution of ARM32 support, as long as (i) it starts at a relatively complete state, i.e., is usable to run wasmtime as-is; and (ii) has an active and interested community of contributors willing to help maintain it and resolve issues.

**Comment by alexcrichton** [comment]
For those following this issue Wasmtime now supports 32-bit ARM via the [Pulley interpreter](https://github.com/bytecodealliance/rfcs/blob/main/accepted/pulley.md) which should now be able to run most wasm programs. Pulley is not 100% finished yet in that we still have some polish to do, but it should be suitable for testing at this time. 

In the meantime I'm going to rename this issue to clarify that the request is to add a Cranelift backend for 32-bit ARM architectures. General support for Wasmtime on 32-bit ARM architectures should be "solved" now and if there are any issues please feel free to open a new issue.

**Comment by mytechnotalent** [comment]
@alexcrichton Just wanted to drop a real-world validation of your recent work on Pulley for 32-bit ARM. I recently got the Pulley interpreter running completely bare-metal on an RP2350 (Cortex-M33) using `#![no_std]` and the Component Model. 

**Implementation details:**
* **Target:** `thumbv8m.main-none-eabihf` (Raspberry Pi Pico 2)
* **Memory:** 256 KiB heap mapped via `embedded-alloc`
* **Execution:** AOT-compiling the guest to `.cwasm` on the host, and interpreting on the MCU.
* **Hardware Abstraction:** Using typed WIT interfaces (`embedded:platform/gpio`) generated by `wit-bindgen` to map directly to the Pico's pins, avoiding raw `env` imports or C-bindings entirely.

The repo and architecture diagram are here if it's a useful reference for anyone else pushing Wasmtime into Cortex-M environments: https://github.com/mytechnotalent/embedded-wasm-blinky

Thanks for pushing the Pulley architecture forward—it makes this kind of bare-metal Wasm actually viable.

**Comment by alexcrichton** [comment]
That's awesome, and thanks for letting us know it's working well for you too! If you spot anything that needs improvement don't hesitate to file an issue

**Comment by mytechnotalent** [comment]
Will do @alexcrichton and thank you for the kind words! I will be building out a large series of drivers and will be testing them and if I see any issues, I will touch base as I am excited to help expand Wasm into the Embedded community!
